src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -7,18 +7,144 @@
+            return;
+        }
+        int hash = hash(key, table.length);
+        if (table[hash] == null) {
+            table[hash] = new Entry<K, V>(key, value, null);
+            size++;
+            return;
+        }
+        Entry<K, V> entry = table[hash];
+        while (entry != null) {
+            if (key.equals(entry.key)) {
+                entry.value = value;
+                return;
+            }
+            entry = entry.next;
+        }
+        addEntry(hash, key, value);
+    }
 
+    private void addEntry(int hash, K key, V value) {
+        table[hash] = new Entry<K, V>(key, value, table[hash]);
+        size++;
+    }
+
+    private void putForNullKey(V value) {
+        Entry<K, V> entry = table[0];
+        while (entry != null) {
+            if (entry.key == null) {
+                entry.value = value;
+                return;
+            }
+            entry = entry.next;
+        }
+        addEntry(0, null, value);
     }
 
     @Override
     public V getValue(K key) {
+        int hash = hash(key, table.length);
+        Entry<K, V> entry = table[hash];
+        while (entry != null) {
+            if (entry.key == null && key == null
+                    || entry.key != null && entry.key.equals(key)) {
+                return entry.value;
+            }
+            entry = entry.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public V remove(K key) {
+        int hash = hash(key, table.length);
+        Entry<K, V> entry = table[hash];
+        Entry<K, V> previousEntry = entry;
+        int counterEntry = 0;
+        while (entry != null) {
+            counterEntry++;
+            if (entry.key == null && key == null
+                    || entry.key != null && entry.key.equals(key)) {
+                if (counterEntry == 1) {
+                    table[hash] = entry.next;
+                }
+                previousEntry.next = entry.next;
+                size--;
+                return entry.value;
+            }
+            previousEntry = entry;
+            entry = entry.next;
+        }
+        return null;
+    }
+
+    private void resize() {
+        Entry<K, V>[] newTable = new Entry[table.length * 2];
+        for (Entry<K, V> entry : table) {
+            while (entry != null) {
+                int hash = hash(entry.key, newTable.length);
+                if (newTable[hash] == null) {
+                    newTable[hash] = new Entry<K, V>(entry.key, entry.value, null);
+                    entry = entry.next;
+                    continue;
+                }
+                newTable[hash] = new Entry<K, V>(entry.key, entry.value, newTable[hash]);
+                entry = entry.next;
+            }
+        }
+        table = newTable;
+    }
+
+    private int hash(K kay, int lengthTable) {
```

Comment:
```suggestion
    private int hash(K key) {
```
we always have access to table.length inside this class

Reply:
yes, but we use this method hash() in method resize() where we must transmit length new table.

Reply:
you always have access to table.length cause you are inside class

Reply:
let's modify this method so it will return 0 for key == null and then you can get rid of putForNullKey

Reply:
done